api/policies/hasMoveType.js

```diff
@@ -0,0 +1,9 @@
+const moveType = require('../../utils/MoveType.json');
+
+module.exports = function (req, res, next) {
+  if (Object.keys(moveType).some((key) => moveType[key] === req.body.moveType)) {
```

Comment:
This creates an array from an object then iterates through the keys, when looking up the key in an object is already a constant time operation.

Three constant time alternatives:

```suggestion
  if (req.body.moveType in MoveType)) {
```


```suggestion
  if (Object.hasOwn(MoveType, req.body.moveType) {
```


```suggestion
  if (MoveType[req.body.moveType]) {
```

Honestly I'm not sure which is best, but I think any of these could be better. I think `in` is the clearest, though it has the theoretical disadvantage of "counting" any properties in the prototype chain as opposed to ones defined in the obj itself, but honestly I think that's fine

Reply:
@seriouslysean curious if you have thoughts on the best way to check an obj for a key

Reply:
Out of those options, I personally prefer `MoveType[req.body.moveType]`. It’s O(1) complexity which means it’s a constant no matter how many items are in the collection. Even if it’s not going to be a large collection we want things to be as efficient as possible.

Additionally, with the loop the way you have it you’re looking at O(n) complexity, which even though isn’t bad it’s less efficient the more items in the array.

That said, the in operator is the fastest since it’s only checking the existence and not actually accessing the property directly. See https://jsbench.me/rim04kjqvx/1.

Either way, always prefer direct key access/validation if possible.

Reply:
So the problem here is that the value will be in the gamestaterow, not the key

Reply:
Is there any way to better align them? Short of that you can make a mapping to lookup the key name, or send it along with the body. Any of those would be better than a loop in my opinion.

Reply:
I didn't realize initially that the issue was in needing a reverse-mapping. We could update the json values to equal the keys so they could be used interchangeably and then update the socket event handlers + all legacy endpoints to match. Then this becomes a dictionary lookup. Honestly it's such a small juice to squeeze ratio that I'm fine with iterating as long as it's more succinct. 

It should be `.values().includes()`

Reply:
I think creating an inverse mapping variable feels a little awkward because we have to keep them in sync. And programmatically generating an inverse mapping variable requires looping anyway. So imo this is okay